### ISSUE TITLE
How to get the local Explanation of  Model as these are not supported by ELI5,LIME,SHAP,etc.

### ISSUE THREAD

**Issue by shashankc28** [fairlearn/fairlearn#818]

**Comment by riedgar-ms** [comment]
Can you elaborate on your question a bit? Fairlearn doesn't provide explanations - you may want to look at:
https://github.com/interpretml/interpret-community

**Comment by shashankc28** [comment]
we are using thresholdoptimizer to mitigate unfairness and we are tring to explain the model which we getting using thressholdoptimzer but this is not supported by lime/shap/eli5....however we tried gridsearch also provided by fairlearn but this is not improving the results.....

**Comment by riedgar-ms** [comment]
I'm not surprised that `ThresholdOptimizer` doesn't work well with a lot of explainers. Its `fit()` method:

- Requires the sensitive attribute to be passed separately, meaning it doesn't quite match the `Estimator` pattern
- Is non-deterministic, which is likely to confuse any explainer

Can you elaborate what you mean  when you say that `GridSearch` is 'not improving the results' since that could mean different things?

**Comment by romanlutz** [comment]
@shashankc28 can you elaborate on what errors you're facing, and which relevant packages and package versions you have installed?

**Comment by shashankc28** [comment]
I am just sharing the sample implementation.
https://colab.research.google.com/drive/1sEn8pX7jTnf-eR2lxjHVlYgvsu8aYeQN?usp=sharing

if possible, request you to please update the notebook itself on how can we improve results and can get explanations.

**Comment by vishals2911** [comment]
@romanlutz , @hildeweerts , @riedgar-ms Thanks team for your prompt responses...am also facing some kind of similar issues....please suggest how can we use XAI libraries with the model produced by Fairlearn threshold optimizer.

**Comment by romanlutz** [comment]
As @hildeweerts explained above there's no way around providing sensitive features for `ThresholdOptimizer`. I would go a step further and say that you shouldn't use it for any real-world application since it's mostly just a benchmark of how much you have to give up in terms of performance to reach parity in terms of a chosen metric across groups. That comes with a few quirks such as applying randomness to outputs. 

What does explaining a random output even mean? Let's take a model with 10% random outputs, 40% depends on the rule `>0.3` and 50% on `>0.7`. Looking inside this model I can tell you that by sheer chance any sample regardless of features can get 0 or 1. If you try it 100 times you'll see how the results will vary and include both 0s and 1s (well, they probably will).
You don't need an explainer here, because you already know exactly how you get to the output. If you have a score less than 0.3 then you have a 10% chance of getting 1, if your score is >0.3 and <0.7 then you have a 50% chance, and if it's >0.7 then you are guaranteed to get a score of 1. Note that some groups will probably not have the baseline 10%. The PR #614 explains this quite nicely I think, but it requires a little more work to get merged, so you won't find that content in the docs yet.
I would always suggest derandomizing these thresholds as much as possible if you're actually considering using something like this. There's also the issue that it's quite obvious that you're using different thresholds per group, so depending on the laws governing your application domain this may be illegal (even though it may improve disparity metrics). I'll also note that fairness is sociotechnical, so there is no one-size-fits-all solution to fairness that can always be expressed through mathematical definitions.

Applying a post-hoc explainer will mislead here since it'll try and figure out how this solution was reached based on random outputs. Instead, the more interesting part would be to explain where your scores came from, since everything after that is quite interpretable as is.

Taking another step back, similar points apply to `fairlearn.reductions.ExponentiatedGradient` since it's also probabilistic. `GridSearch` should be just fine because it decides on a single predictor internally rather than probabilistically choosing one of many. Additionally, neither of the reductions techniques require `sensitive_features` at `predict` time.

I don't quite understand your concern about `GridSearch` in the notebook. This is your code:
```
sweep = GridSearch(RandomForestClassifier(n_estimators=30, random_state=0),
                   constraints=DemographicParity(),
                   grid_size=71)
sweep.fit(X_train, y_train,
          sensitive_features=x_gender)
```
Later you use existing code to calculate non-dominated models, but comment it out to just get all models generated from the grid, and you take the first one of those to create predictions:
```
y_pred_mitigated=non_dominated[0].predict(X_test)
```
which raises the question: why the first one? It's very likely not the best model you can choose. If you want the default choice just go with the `sweep` object and call `predict` on it. All this extra code looks like it's been copied from a notebook of ours and is mainly there to allow plotting to illustrate the fairness-accuracy trade-off, but since you're not doing that I don't see why you'd choose that model manually (as model 0). It's always recommended to factor in that trade-off (potentially in more than 2 dimensions) when choosing a model.

Later you write
> Now trying eli5 on model selected by gridsearch it works fine for explanation but model is not perfect as we have seen selection rate above.
That is probably related to choosing model 0 rather than one with a better trade-off. You can always plot the selection rate vs. another metric for all the predictors internal to grid search if you want to use that as a criterion for selecting a model.

All in all, the models resulting from `GridSearch` should work with post-hoc explainers.

**Comment by romanlutz** [comment]
Closing this issue for now. If there are further questions please reopen. We can also consider explicitly supporting compatibility with some explainers, although I would discourage using blackbox explainers for some of our mitigation techniques as explained above.
